### PR TITLE
ARROW-13164: [R] altrep vectors from Array with nulls

### PR DESCRIPTION
**Pull request by romainfrancois**

### CODE REVIEW

**Comment by romainfrancois** [comment]
follow up after https://github.com/apache/arrow/pull/10593

File: r/src/array_to_vector.cpp [hunk 1]
```diff
@@ -69,13 +69,25 @@ class Converter {
...
-          array->null_count() == 0) {
+      // using altrep if
+      // - the arrow.use_altrep is set to TRUE or unset
+      // - the array has at least one element
+      // - either it has no nulls or the data is mutable
+      if (arrow::r::GetBoolOption("arrow.use_altrep", true) && array->length() > 0) {
         switch (array->type()->id()) {
           case arrow::Type::DOUBLE:
-            return arrow::r::MakeDoubleArrayNoNull(array);
+            if (array->null_count() == 0 || array->data()->buffers[1]->is_mutable()) {
```
Comment (bkietz) [inline r/src/array_to_vector.cpp:79]: It's not safe to do this if anyone else might mutate this buffer. We don't have a completely safe way to check uniqueness, either. As a stop gap:
```suggestion
            if (array->null_count() == 0 || array->data().use_count() == 1 &&
                array->data()->buffers[1].use_count() == 1 && array->data()->buffers[1]->is_mutable()) {
```

File: r/src/altrep.cpp [hunk 2]
```diff
@@ -44,25 +44,68 @@ extern "C" {
...
-    // that retain the array
+    // that retain the Array
     Pointer xp(new std::shared_ptr<Array>(array));
 
+    // we only get here if the Array data buffer is mutable
+    // UseSentinel() puts the R sentinel where the data is null
+    auto null_count = array->null_count();
+    if (null_count > 0) {
+      tasks.Append(true, [array]() {
+        UseSentinel<data_type>(array);
```
Comment (bkietz) [inline r/src/altrep.cpp:104]: In the case of a few long arrays, it'd be more performant to divide this work into L2 sized chunks

File: r/src/altrep.cpp [hunk 3]
```diff
@@ -44,25 +44,68 @@ extern "C" {
...
+                   RTasks& tasks) {
     // we don't need the whole r6 object, just an external pointer
-    // that retain the array
+    // that retain the Array
     Pointer xp(new std::shared_ptr<Array>(array));
 
+    // we only get here if the Array data buffer is mutable
+    // UseSentinel() puts the R sentinel where the data is null
+    auto null_count = array->null_count();
+    if (null_count > 0) {
```
Comment (bkietz) [inline r/src/altrep.cpp:102]: Would it make sense to do this lazily in dataptr/getregion?

File: r/src/altrep.cpp [hunk 4]
```diff
@@ -44,25 +44,68 @@ extern "C" {
...
 #include <arrow/array.h>
+#include <arrow/util/bitmap_reader.h>
+
+#include "./r_task_group.h"
 
 namespace arrow {
 namespace r {
 
+template <typename T>
+T na_sentinel();
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply r/src/altrep.cpp]: We can :-)

File: r/src/altrep.cpp (hunk 3 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:97]: maybe, but once we're in Dataptr/Getregion we no longer have the possibility to do this in a task. 

Here, we always do it (although we might not need to), but in a task (so in parallel). 
If we did it in Dataptr/Getregion we would have to bookkeep if it was done already and if not do it serially.

File: r/src/altrep.cpp (hunk 2 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:99]: Not sure how to do this. Maybe this can be a follow up issue.

**Comment by romainfrancois** [comment]
this needs tests about min/max/sum

File: r/src/array_to_vector.cpp (hunk 1 above)
Comment (bkietz) [reply r/src/array_to_vector.cpp]: @pitrou

Comment (romainfrancois) [reply r/src/array_to_vector.cpp]: One consequence is that we can't use altrep when taking slices of an Array as the buffer is at least held by two Arrays then. 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

is_altrep <- arrow:::is_altrep

v_int <- Array$create(c(1L, NA, 3L))

is_altrep(v_int$as_vector())
#> [1] TRUE

is_altrep(as.vector(v_int$Slice(1)))
#> [1] FALSE
```

<sup>Created on 2021-07-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

I commented out the relevant tests in `test-altrep.R`. Does this matter given that the data there is irrelevant because it's identified as null ?

File: r/tests/testthat/test-altrep.R [hunk 5]
```diff
@@ -59,52 +59,117 @@ test_that("altrep vectors from int32 and dbl arrays with nulls", {
...
   df_batch <- as.data.frame(batch)
-  expect_true(is_altrep_int_nonull(df_batch$int))
-  expect_true(is_altrep_dbl_nonull(df_batch$dbl))
+  expect_true(is_altrep(df_batch$int))
+  expect_true(is_altrep(df_batch$dbl))
+})
+
+test_that("altrep min/max/sum identical to R versions for double", {
+  expect_altrep_rountrip <- function(x, fn, ...) {
+    expect_identical(fn(x, ...), fn(Array$create(x)$as_vector(), ...))
```
Comment (nealrichardson) [inline r/tests/testthat/test-altrep.R:113]: Should this also assert altrep-ness, maybe something like this?

```suggestion
    alt <- Array$create(x)$as_vector()
    expect_true(is_altrep(alt))
    result <- fn(alt)
    # Calling fn() didn't cause the vector to lose its altrep-ness
    expect_true(is_altrep(alt))
    expect_identical(result, fn(x, ...), ...))
```

File: r/src/array_to_vector.cpp [hunk 6]
```diff
@@ -133,6 +146,16 @@ class Converter {
...
   std::shared_ptr<ChunkedArray> chunked_array_;
+
+ private:
+  bool CanAltrep(const std::shared_ptr<Array>& array) {
+    if (array->null_count() == 0) {
+      return true;
+    }
+
+    return array->data().use_count() == 1 && array->data()->buffers[1].use_count() == 1 &&
+           array->data()->buffers[1]->is_mutable();
```
Comment (nealrichardson) [inline r/src/array_to_vector.cpp:157]: I thought arrays were immutable? when/how likely is this to be true?

File: r/src/altrep.cpp [hunk 7]
```diff
@@ -104,63 +143,138 @@ struct ArrayNoNull {
...
+    auto options = Options(array, na_rm);
+
+    const auto& sum =
+        ValueOrStop(arrow::compute::CallFunction("sum", {array}, options.get()));
+
+    if (sexp_type == INTSXP) {
+      // When calling the "sum" function on an int32 array, we get an Int64 scalar
+      // in case of overflow, make it a double like R
+      int64_t value = internal::checked_cast<const Int64Scalar&>(*sum.scalar()).value;
+      if (value < INT32_MIN || value > INT32_MAX) {
```
Comment (nealrichardson) [inline r/src/altrep.cpp:193]: `INT32_MIN` is `NA_integer_` right? So it needs to be out of bounds too. Or does this constant come from R and already have that factored in?

```suggestion
      if (value <= INT32_MIN || value > INT32_MAX) {
```

File: r/src/altrep.cpp [hunk 8]
```diff
@@ -44,25 +49,58 @@ extern "C" {
...
 #include <arrow/array.h>
+#include <arrow/util/bitmap_reader.h>
+
+#include "./r_task_group.h"
 
 namespace arrow {
 namespace r {
 
+template <typename T>
+void UseSentinel(const std::shared_ptr<Array>& array) {
```
Comment (nealrichardson) [inline r/src/altrep.cpp:60]: I would have expected that, instead of altering the Arrow data, we would have `_ELT` and `_GET_REGION` methods that would insert the sentinels in the R vector when the data is accessed. What are the tradeoffs of that way vs. this way?

File: r/src/array_to_vector.cpp (hunk 6 above)
Comment (romainfrancois) [reply r/src/array_to_vector.cpp:157]: specifically here when mutating does not logically change anything. it's my understanding that what's there is undefined and that it is inconsequential to change it. 

it may be false however when the buffer comes from a source that prevents mutation, I don't have a specific example but I guess reading from disk or something.

File: r/tests/testthat/test-altrep.R (hunk 5 above)
Comment (romainfrancois) [reply r/tests/testthat/test-altrep.R:113]: Thanks. I'll update to add more things to `expect_altrep_rountrip()`

File: r/src/altrep.cpp (hunk 7 above)
Comment (romainfrancois) [reply r/src/altrep.cpp]: Thanks. R indeed converts to numeric in that case: 

``` r
str(sum(as.integer(c(-2^31 + 1L))))
#>  int -2147483647
str(sum(as.integer(c(-2^31 + 1L, -1L))))
#>  num -2.15e+09
```

<sup>Created on 2021-07-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

File: r/src/altrep.cpp (hunk 8 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:60]: Oh interesting. I guess the advantage of this approach is that, for the price of this initial setting of the sentinels, we can then have a `DATAPTR()` so access all the data of the array randomly. 

OTOH, we could leave the data untouched and have an `_Elt` that would check the null bitmap, and an `GetRegion()` that would also locally use the null bitmap. This `GetRegion()` would indeed copy the region data though, so potentially be more expensive, even when the region has no nulls.

Comment (romainfrancois) [reply r/src/altrep.cpp:60]: Chances are we might need to implement something like this for arrays that are backed by an immutable buffer, so maybe I can do that, and then we can compare/measure and evaluate how risky this is. 

Then we have 3 cases: 
 - Array with no nulls: best altrep with DATAPTR(), no need for setting sentinels (what we had prior to this pull request). 
 - Array with some nulls, immutable: using GetRegion, I guess with a materialized R vector in the data2 altrep slot, so that DATAPTR() materializes and returns this. 
 
 - Array with some nulls, "mutable": hacky proposal of this pr. 

Perhaps we only need the first 2.

Comment (romainfrancois) [reply r/src/altrep.cpp:60]: Also, I believe the second dot will have some code shared with altrep for ChunkedArray.

File: r/tests/testthat/test-altrep.R [hunk 9]
```diff
@@ -59,52 +59,121 @@ test_that("altrep vectors from int32 and dbl arrays with nulls", {
...
+  expect_altrep_rountrip(x, max, na.rm = TRUE)
+  expect_altrep_rountrip(x, sum, na.rm = TRUE)
+
+  expect_altrep_rountrip(x, min)
+  expect_altrep_rountrip(x, max)
+  expect_altrep_rountrip(x, sum)
+
+  x <- rep(NA_real_, 3)
+  expect_warning(expect_altrep_rountrip(x, min, na.rm = TRUE))
+  expect_warning(expect_altrep_rountrip(x, max, na.rm = TRUE))
```
Comment (pitrou) [inline r/tests/testthat/test-altrep.R:140]: Is there a way to test the precise warning that is being emitted here? (or at least a substring thereof)

File: r/tests/testthat/test-altrep.R [hunk 10]
```diff
@@ -59,52 +59,121 @@ test_that("altrep vectors from int32 and dbl arrays with nulls", {
...
+  expect_altrep_rountrip(x, max, na.rm = TRUE)
+  expect_altrep_rountrip(x, sum, na.rm = TRUE)
+
+  expect_altrep_rountrip(x, min)
+  expect_altrep_rountrip(x, max)
+  expect_altrep_rountrip(x, sum)
+
+  x <- rep(NA_integer_, 3)
+  expect_warning(expect_altrep_rountrip(x, min, na.rm = TRUE))
+  expect_warning(expect_altrep_rountrip(x, max, na.rm = TRUE))
```
Comment (pitrou) [inline r/tests/testthat/test-altrep.R:169]: Same here.

File: r/src/array_to_vector.cpp (hunk 6 above)
Comment (pitrou) [reply r/src/array_to_vector.cpp:157]: Hmm, yes, I guess it _should_ be logicall inconsequential. OTOH, it may come as a surprise if people e.g. blindlessly memory-map a file in read-write mode without imagining that RArrow will modify it.

File: r/src/array_to_vector.cpp [hunk 11]
```diff
@@ -133,6 +146,16 @@ class Converter {
...
  protected:
   std::shared_ptr<ChunkedArray> chunked_array_;
+
+ private:
+  bool CanAltrep(const std::shared_ptr<Array>& array) {
+    if (array->null_count() == 0) {
+      return true;
+    }
+
+    return array->data().use_count() == 1 && array->data()->buffers[1].use_count() == 1 &&
```
Comment (pitrou) [inline r/src/array_to_vector.cpp:156]: Why the `use_count` check? Even if the Buffer _object_ is unshared, the underlying _data_ may be shared (for example because a slice was made of a parent Buffer).

File: r/src/altrep.cpp [hunk 12]
```diff
@@ -19,9 +19,14 @@
 
 #if defined(ARROW_R_WITH_ARROW)
 
+#include <arrow/compute/api.h>
```
Comment (pitrou) [inline r/src/altrep.cpp:22]: Would be nice if you could cluster all Arrow includes together.

File: r/src/altrep.cpp [hunk 13]
```diff
@@ -104,63 +143,138 @@ struct ArrayNoNull {
...
 void Init_Altrep_classes(DllInfo* dll) {
-  DoubleArrayNoNull::Init(dll);
-  Int32ArrayNoNull::Init(dll);
+  AltrepVectorDouble::Init(dll);
+  AltrepVectorInt32::Init(dll);
 }
 
-SEXP MakeDoubleArrayNoNull(const std::shared_ptr<Array>& array) {
-  return DoubleArrayNoNull::Make(array);
+SEXP MakeAltrepVectorDouble(const std::shared_ptr<Array>& array, RTasks& tasks) {
```
Comment (pitrou) [inline r/src/altrep.cpp:272]: Since `tasks` is a mutable argument, can you make it `RTasks* tasks`?

File: r/src/array_to_vector.cpp (hunk 11 above)
Comment (romainfrancois) [reply r/src/array_to_vector.cpp:147]: This is no longer being used.

**Comment by romainfrancois** [comment]
Thanks for the reviews. I'll apply the relevant changes. 

We now have `AltrepArrayNoNulls<INTSXP|REALSXP>` and `AltrepArrayWithNulls<INTSXP|REALSXP>` which are not all that different. I believe they can be merged into one template, which would perhaps slightly more involved `Dataptr()` ... methods...

File: r/src/altrep.cpp (hunk 13 above)
Comment (romainfrancois) [reply r/src/altrep.cpp]: we no longer use tasks there as there is nothing to parallelise

**Comment by romainfrancois** [comment]
I finally settled for a single altrep (template) class for both cases (the array has no nulls or the array has nulls). 

The altrep object uses: 
 - data1: an external pointer to the Array (well to a shared pointer to the Array) 
 - data2: NULL until materialization is needed, then data2 is a standard (and immutable) R vector with the same data

Materialization is needed when: 
 - we need the DATAPTR and the Array has nulls
 - we need a duplicate

Get_region() does not need to materialize the full vector.
Creating the altrep object does not alter the data to impose the R sentinel.

File: r/src/array_to_vector.cpp [hunk 14]
```diff
@@ -133,6 +131,16 @@ class Converter {
 
  protected:
   std::shared_ptr<ChunkedArray> chunked_array_;
+
+ private:
+  bool CanAltrep(const std::shared_ptr<Array>& array) {
```
Comment (pitrou) [inline r/src/array_to_vector.cpp:136]: Can you remove this method?

File: r/src/altrep.cpp [hunk 15]
```diff
@@ -43,144 +50,356 @@ extern "C" {
...
+    return array_->data()->template GetValues<data_type>(1);
   }
 
-  static SEXP Make(const std::shared_ptr<Array>& array) {
-    return ArrayNoNull<INTSXP>::Make(class_t, array);
+  data_type* array_data() const {
+    return const_cast<data_type*>(array_->data()->template GetValues<data_type>(1));
+  }
+
+  R_xlen_t Get_region(R_xlen_t i, R_xlen_t n, data_type* buf) {
```
Comment (pitrou) [inline r/src/altrep.cpp:168]: 1) Can you add comments describing non-trivial methods or functions?
2) Can you try to follow the naming conventions (here `GetRegion`)?

File: r/src/altrep.cpp [hunk 16]
```diff
@@ -43,144 +50,356 @@ extern "C" {
...
-    R_set_altvec_Dataptr_method(class_t, ArrayNoNull::Dataptr);
-    R_set_altvec_Dataptr_or_null_method(class_t, ArrayNoNull::Dataptr_or_null);
+    return NULL;
   }
-};
 
-struct DoubleArrayNoNull {
-  static R_altrep_class_t class_t;
+  // force materialization, and then return data ptr
+  void* Dataptr(Rboolean writeable) {
```
Comment (pitrou) [inline r/src/altrep.cpp:138]: Why do we have `Dataptr`, `DATAPTR`, `data1`, `data2`... can you try to make the names clearer and/or add comments explaining the intent?

File: r/src/altrep.cpp [hunk 17]
```diff
@@ -43,144 +50,356 @@ extern "C" {
...
+    if (array_->null_count() == 0) {
+      return reinterpret_cast<void*>(array_data());
```
Comment (pitrou) [inline r/src/altrep.cpp:140]: Does `writeable` mean the user wants to write to the data? If so, we should not return the original Arrow data in this case.

File: r/src/altrep.cpp [hunk 18]
```diff
@@ -43,144 +50,356 @@ extern "C" {
...
   }
 
-  static SEXP Make(const std::shared_ptr<Array>& array) {
-    return ArrayNoNull<INTSXP>::Make(class_t, array);
+  data_type* array_data() const {
+    return const_cast<data_type*>(array_->data()->template GetValues<data_type>(1));
+  }
+
+  R_xlen_t Get_region(R_xlen_t i, R_xlen_t n, data_type* buf) {
+    const auto& slice = array_->Slice(i, n);
```
Comment (pitrou) [inline r/src/altrep.cpp:169]: I don't think you can take a const reference here. `Array::Slice` is returning a new `shared_ptr`.

File: r/src/altrep.cpp [hunk 19]
```diff
@@ -43,144 +50,356 @@ extern "C" {
...
-  // altrep object around an Array with no nulls
-  // data1: an external pointer to a shared pointer to the Array
-  // data2: not used
+  static R_altrep_class_t class_t;
 
-  static SEXP Make(R_altrep_class_t class_t, const std::shared_ptr<Array>& array) {
-    // we don't need the whole r6 object, just an external pointer
-    // that retain the array
-    Pointer xp(new std::shared_ptr<Array>(array));
+  using data_type = typename std::conditional<sexp_type == REALSXP, double, int>::type;
```
Comment (pitrou) [inline r/src/altrep.cpp:68]: Nit: call this `c_type`?

File: r/src/altrep.cpp [hunk 20]
```diff
@@ -43,144 +50,356 @@ extern "C" {
...
+
+  R_xlen_t Get_region(R_xlen_t i, R_xlen_t n, data_type* buf) {
+    const auto& slice = array_->Slice(i, n);
+
+    R_xlen_t ncopy = slice->length();
+
+    if (IsMaterialized()) {
+      // just use data2
+      memcpy(buf, reinterpret_cast<data_type*>(DATAPTR(data2())) + i,
+             ncopy * sizeof(data_type));
```
Comment (pitrou) [inline r/src/altrep.cpp:176]: If it's already materialized, why do a copy again? Is there a situation in which this code path is executed?

**Review by pitrou — CHANGES_REQUESTED** [review]
I'm going to say that the `AltrepArrayPrimitive` seems very difficult to read and understand to me.

I would recommend:
1) write extensive comments and explanations (including hyperlinks to R-related docs where necessary)
2) choose naming wisely
3) make sure no unused functionality or code paths are left
4) make internal helper methods private, to better stress what is the "public" API
5) also, make sure we're not doing any logic errors (if the user requests a writable array, do we really want to return a pointer to the immutable Arrow data?)

File: r/src/altrep.cpp (hunk 16 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:138]: Those are the R names, we can't do much about this here.

File: r/src/altrep.cpp (hunk 15 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:168]: Just trying to align with the R conventions used by the functions calling this.

File: r/src/altrep.cpp (hunk 17 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:140]: I need to check that, but yeah probably

File: r/src/altrep.cpp (hunk 20 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:176]: This is what `Get_region` is about: copy the slice from `i` to `i + n` into `buf`

File: r/src/altrep.cpp (hunk 17 above)
Comment (romainfrancois) [reply r/src/altrep.cpp]: In a way yes, but the user should not want that because the object is marked as immutable (in the R sense). Unfortunately, this is more of a guideline than something that is enforced by R. 

Accessing data read-only is somewhat new in R, and lots of internal R code would still call `DATAPTR()` even though a read)-only equivalent would be sufficient.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit -g r

**Comment by nealrichardson** [comment]
I _think_ that's all clean. @jonkeane do you want to do any benchmarking before we merge this?

**Comment by nealrichardson** [comment]
@ursabot please benchmark lang=R

**Comment by jonkeane** [comment]
Yeah, we should run these manually. I'll try to get to that tomorrow. None of the array conversion benchmarks have been wired up in the benchmarks package, so we don't have them available on conbench right now.

**Comment by jonkeane** [comment]
OK, I just ran Array to R benchmarks locally, and the results are very impressive. I've added in other types (even though they are not (yet) backed by altrep). I've also tested both chunked and non-chunked arrays, and with nulls and without.

TL;DR:

For floats + ints we see huge speed ups (~10x faster) in both arrays with and without nulls as well as in arrays and chunked arrays (at least in the case where there is a single chunk*).

I'm still investigating what's going on with the two fannie parquet file reads that show up as regressions. If anything, this should have sped them up. 

* – we don't yet have fixtures with chunked arrays with multiple chunks. But I simulated this with a csv (which does come in with multiple chunks) and multi-chunk arrays aren't (yet) supported:

``` r
library(arrow)

nyctaxi <- read_csv_arrow("~/repos/ab_store/data/nyctaxi_2010-01.csv.gz", as_data_frame = FALSE)

chunked_array <- as.vector(nyctaxi[[5]])
.Internal(inspect(chunked_array))
#> @7fb218000000 14 REALSXP g0c7 [REF(4)] (len=14863778, tl=0) 0.75,5.9,4,4.7,0.6,...

array <- as.vector(nyctaxi[[5]]$chunk(1))
.Internal(inspect(array))
#> @7fb20dc3a6e0 14 REALSXP g0c0 [REF(65535)] arrow::Array<double, NONULL> len=5738, Array=<0x7fb25d5080d8>
#>   @7fb20dc3a670 22 EXTPTRSXP g0c0 [REF(4)]
```

**Comment by nealrichardson** [comment]
> * multi-chunk arrays aren't (yet) supported

That's ARROW-13111. ARROW-13112 is for supporting other types. If I understand correctly, we're conceptually close on those since we're implementing the get_region and elt methods for integer and float here, just ("just"!) need to [something about templates] and do the same for other types and for ChunkedArrays using our existing conversion code.
